File: plugins/Lyrics/src/main/java/com/aliucord/plugins/Lyrics.java
package com.aliucord.plugins;

import android.content.Context;

import androidx.annotation.NonNull;

import com.aliucord.Http;
import com.aliucord.api.CommandsAPI;
import com.aliucord.entities.MessageEmbedBuilder;
import com.aliucord.entities.Plugin;
import com.aliucord.plugins.lyrics.Data;
import com.aliucord.plugins.lyrics.ResponseModel;
import com.aliucord.utils.RxUtils;
import com.discord.api.commands.ApplicationCommandType;
import com.discord.models.commands.ApplicationCommandOption;
import com.discord.models.domain.spotify.ModelSpotifyTrack;
import com.discord.stores.StoreStream;
import com.discord.utilities.spotify.SpotifyApiClient;

import java.util.Arrays;
import java.util.Collections;

import rx.Subscriber;

@SuppressWarnings("unused")
public class Lyrics extends Plugin {

    private static final String baseUrl = "https://lyrics-api.powercord.dev/lyrics?input=";
    private static final int MAX_MESSAGE_LENGTH = 2000;

    @NonNull
    @Override
    public Manifest getManifest() {
        var manifest = new Manifest();
        manifest.authors = new Manifest.Author[] { new Manifest.Author("Xinto",423915768191647755L) };
        manifest.description = "Get lyrics to a specific song.";
        manifest.version = "1.3.0";
        manifest.updateUrl = "https://raw.githubusercontent.com/X1nto/AliucordPlugins/builds/updater.json";
        return manifest;
    }

    @Override
    public void start(Context context) {
        var songNameArg = new ApplicationCommandOption(ApplicationCommandType.STRING, "name", "The song name to search lyrics for", null, false, false, null, null);
        var shouldSendArg = new ApplicationCommandOption(ApplicationCommandType.BOOLEAN, "send", "To send output in the chat or not", null, false, false, null, null);
        var arguments = Arrays.asList(songNameArg, shouldSendArg);

        commands.registerCommand("lyrics", "Grab a song lyrics", arguments, ctx -> {
            var shouldSend = ctx.getBool("send");
            final String[] songName = { ctx.getString("name") };

            if (shouldSend == null) {
                shouldSend = false;
            }

            if (songName[0] == null) {
                try {
                    var storeSpotify = StoreStream.getSpotify();

                    var _spotifyApiClient = storeSpotify.getClass().getDeclaredField("spotifyApiClient");
                    _spotifyApiClient.setAccessible(true);

                    var spotifyApiClient = (SpotifyApiClient) _spotifyApiClient.get(storeSpotify);

                    if (spotifyApiClient == null) {
                        return new CommandsAPI.CommandResult("Failed to get current Spotify activity.", null, false);
                    }

                    //This sucks so much
                    RxUtils.subscribe(spotifyApiClient.getSpotifyTrack(), new Subscriber<>() {
                        @Override
                        public void onCompleted() {}

                        @Override
                        public void onError(Throwable th) {
                            songName[0] = "";
                        }

                        @Override
                        public void onNext(ModelSpotifyTrack modelSpotifyTrack) {
                            songName[0] = String.format("%s %s", modelSpotifyTrack.getArtists().get(0).getName(), modelSpotifyTrack.getName());
                        }
                    });

                } catch (Exception ignored) {}
            }

            //don't fetch anything until songName is not null
            //noinspection StatementWithEmptyBody, LoopConditionNotUpdatedInsideLoop
            while (songName[0] == null) {}

            if (songName[0].equals("")) {
                return new CommandsAPI.CommandResult("Failed to get current Spotify activity.", null, false);
            }

            try {
                var data = fetch(songName[0]);
                return shouldSend ? lyricsText(data) : lyricsEmbed(data);
            } catch (Exception e) {
                e.printStackTrace();
                return new CommandsAPI.CommandResult("Failed to fetch data", null, false);
            }
        });
    }

    @Override
    public void stop(Context context) {
        commands.unregisterAll();
    }

    private CommandsAPI.CommandResult lyricsText(Data data) {
        var lyrics = data.lyrics;

        if (lyrics.length() > MAX_MESSAGE_LENGTH) {
            var fullLyricsText = String.format("\n\nFull Lyrics: %s", data.url);
            lyrics = lyrics.substring(0, MAX_MESSAGE_LENGTH - fullLyricsText.length() - 3) + "..." + fullLyricsText;
        }

        return new CommandsAPI.CommandResult(lyrics);
    }

    private CommandsAPI.CommandResult lyricsEmbed(Data data) {
        var embed = new MessageEmbedBuilder()
                .setAuthor(data.artist, null, null)
                .setTitle(data.name)
                .setDescription(data.lyrics)
                .setUrl(data.url)
                .setThumbnail(data.album_art)
                .setColor(0x209CEE)
                .setFooter(String.format("Lyrics provided by KSoft.Si | © %s %s", data.artist, data.album_year.split(",")[0]), "https://external-content.duckduckgo.com/iu/?u=https://cdn.ksoft.si/images/Logo128.png")
                .build();

        return new CommandsAPI.CommandResult(null, Collections.singletonList(embed), false, "Lyrics");
    }

    private Data fetch(String song) throws Exception {
        var responseModel = (ResponseModel) Http.simpleJsonGet(baseUrl + song, ResponseModel.class);
        return responseModel.data.get(0);
    }

}

File: plugins/Lyrics/src/main/java/com/aliucord/plugins/Lyrics.kt
package com.aliucord.plugins

import android.content.Context
import com.aliucord.Http
import com.aliucord.Utils
import com.aliucord.annotations.AliucordPlugin
import com.aliucord.api.CommandsAPI.CommandResult
import com.aliucord.entities.CommandContext
import com.aliucord.entities.MessageEmbedBuilder
import com.aliucord.entities.Plugin
import com.aliucord.plugins.lyrics.Data
import com.aliucord.plugins.lyrics.ResponseModel
import com.aliucord.utils.RxUtils.subscribe
import com.discord.api.commands.ApplicationCommandType
import com.discord.models.domain.spotify.ModelSpotifyTrack
import com.discord.stores.StoreStream
import com.discord.utilities.spotify.SpotifyApiClient
import rx.Subscriber
import java.util.*

@AliucordPlugin
class Lyrics : Plugin() {

    private val baseUrl = "https://lyrics-api.powercord.dev/lyrics?input="
    private val maxMessageLength = 2000

    override fun start(context: Context) {
        val songNameArg = Utils.createCommandOption(
            type = ApplicationCommandType.STRING,
            name = "name",
            description = "The song name to search lyrics for",
        )

        val shouldSendArg = Utils.createCommandOption(
            type = ApplicationCommandType.STRING,
            name = "send",
            description = "Whether to send the lyrics to chat",
        )

        commands.registerCommand(
            "lyrics",
            "Grab a song lyrics",
            listOf(songNameArg, shouldSendArg)
        ) { ctx: CommandContext ->
            val shouldSend = ctx.getBool("send") ?: false
            var songName = ctx.getString("name")

            if (songName == null) {
                val storeSpotify = StoreStream.getSpotify()
                val spotifyApiClient = storeSpotify.javaClass
                    .getDeclaredField("spotifyApiClient")
                    .let {
                        it.isAccessible = true
                        it.get(storeSpotify) as SpotifyApiClient
                    }

                //This sucks so much
                spotifyApiClient.spotifyTrack.subscribe(
                    object : Subscriber<ModelSpotifyTrack?>() {
                        override fun onCompleted() {}
                        override fun onError(th: Throwable) {
                            songName = ""
                        }

                        override fun onNext(modelSpotifyTrack: ModelSpotifyTrack?) {
                            if (modelSpotifyTrack == null) return
                            songName = "${modelSpotifyTrack.artists[0].name} ${modelSpotifyTrack.name}"
                        }
                    }
                )
            }

            //don't fetch anything until songName is not null
            @Suppress("ControlFlowWithEmptyBody")
            while (songName == null) {}

            if (songName == "") {
                return@registerCommand CommandResult(
                    "Failed to get current Spotify activity.",
                    null,
                    false
                )
            }

            val data = fetch(songName)
            return@registerCommand if (shouldSend) lyricsText(data) else lyricsEmbed(data)
        }
    }

    override fun stop(context: Context) {
        commands.unregisterAll()
    }

    private fun lyricsText(data: Data): CommandResult {
        var lyrics = data.lyrics
        if (lyrics.length > maxMessageLength) {
            val fullLyricsText = "\n\nFull Lyrics: ${data.url}"
            lyrics = lyrics.substring(0, maxMessageLength - fullLyricsText.length - 3) + "..." + fullLyricsText
        }
        return CommandResult(lyrics)
    }

    private fun lyricsEmbed(data: Data): CommandResult {
        val embed = MessageEmbedBuilder()
            .setAuthor(data.artist, null, null)
            .setTitle(data.name)
            .setDescription(data.lyrics)
            .setUrl(data.url)
            .setThumbnail(data.album_art)
            .setColor(0x209CEE)
            .setFooter(
                "Lyrics provided by KSoft.Si | © ${data.artist} ${data.album_year.split(",")[0]}",
                "https://external-content.duckduckgo.com/iu/?u=https://cdn.ksoft.si/images/Logo128.png"
            )
            .build()
        return CommandResult(null, listOf(embed), false, "Lyrics")
    }

    private fun fetch(song: String?): Data {
        val responseModel = Http.simpleJsonGet(baseUrl + song, ResponseModel::class.java) as ResponseModel
        return responseModel.data[0]
    }

}